File: Array/658.Find_K_Closest_Elements/Python/658.Find_K_Closest_Elements.py
class Solution(object):
    def findClosestElements(self, arr, k, x):
        """
        :type arr: List[int]
        :type k: int
        :type x: int
        :rtype: List[int]
        """
        #=====================================#
        # Two-pointers based traversal method #
        #=====================================#
        if (x >= arr[(-1)]): #Issue/Boundary-case handling (i.e. exceed maximun)
            return arr[(len(arr) - k): ]

        elif (x <= arr[0]): #Issue/Boundary-case handling (i.e. below minimun)
            return arr[: k]

        ############
        #Initialize
        ##### Length of arr array #####
        len_arr = len(arr)

        ##### Record two-pointers (i.e. left, right) #####
        record_left_ptr, record_right_ptr = 0, 1

        ##### Result array #####
        res_arr = []


        ###################################
        #Two-pointers based loop traversal
        ##### Step 1: Looped-traversal for the index/place of x value #####
        while (record_left_ptr < len_arr):

            ##### Check if the current indexed-value matched conditions or not #####
            if (arr[(record_left_ptr + 1)] > x):
                record_right_ptr += record_left_ptr #Update/Accumulate

                break

            record_left_ptr += 1 #Keep updating/accumulating

        ##### Step 2: Post-process/Final completions #####
        for _ in range(k):

            ##### Check if the current pointer exceed boundary or not #####
            if (record_left_ptr >= 0):
                if (record_right_ptr < len_arr):
                    curr_left_diff = (x - arr[record_left_ptr]) #Current left-half indexed-value differences
                    curr_right_diff = (arr[record_right_ptr] - x) #Current right-half indexed-value differences
                
                    ##### Check if the current indexed-value matched conditions or not #####
                    if (curr_left_diff <= curr_right_diff):
                        res_arr.insert(0, arr[record_left_ptr]) #Keep updating/recording

                        record_left_ptr -= 1 #Keep updating/reducing

                    else:
                        res_arr.append(arr[record_right_ptr]) #Keep updating/recording

                        record_right_ptr += 1 #Keep updating/accumulating

                else:
                    res_arr.insert(0, arr[record_left_ptr]) #Keep updating/recording

                    record_left_ptr -= 1 #Keep updating/reducing

            else:
                ##### Check if the current pointer exceed boundary or not #####
                if (record_right_ptr < len_arr):
                    res_arr.append(arr[record_right_ptr]) #Keep updating/recording

                    record_right_ptr += 1 #Keep updating/accumulating

                else:
                    pass

        return res_arr

File: Array/658.Find_K_Closest_Elements/C++/658.Find_K_Closest_Elements.cpp
class Solution {
public:
    vector<int> findClosestElements(vector<int> & arr, int k, int x) 
    {
        //===================================//
        // Input type:                       //
        //   - vector<int> ref. (arr vector) //
        //   - int (k-th elements)           //
        //   - int (x target value)          //
        // Return type:                      //
        //   - vector<int> (result vector)   //
        //===================================//
        if (x <= arr.front()) //Issue, Boundary-case handling
        {
            vector<int> sub_vect((arr.begin() + 0), (arr.begin() + k)); //Sub-vector from 0 to k-indexed

            return sub_vect; 
        }

        else if (x >= arr.back()) //Issue, Boundary-case handling
        {
            vector<int> sub_vect((arr.begin() + (arr.size() - k)), (arr.begin() + arr.size())); //Sub-vector from the last k-th elements/values

            return sub_vect;
        }


        /*Initialize*/
        ///// Length of arr vector /////
        int len_arr = arr.size();

        ///// Record two-pointers (i.e. left, right) /////
        int record_left_ptr = 0, record_right_ptr = 1;

        ///// Record differences (i.e. left, right) /////
        int record_left_diff, record_right_diff;

        ///// Result vector /////
        vector<int> res_vect;


        /*Two-pointers based loop traversal*/
        ///// Step 1: Looped-traversal for the index/place of x target value /////
        while (record_left_ptr < len_arr) //Part
        {
            ///// Check if the current indexed-value matched conditions or not /////
            if (arr[(record_left_ptr + 1)] > x)
            {
                record_right_ptr += record_left_ptr; //Update, Accumulate

                break;
            }

            (record_left_ptr++); //Keep updating, accumulating

        } //Part

        ///// Step 2: Post-process, Final completions /////
        for (int k_idx = 0; (k_idx < k); (k_idx++)) //K-th times
        {
            ///// Check if the current pointer exceed boundary or not /////
            if (record_left_ptr >= 0)
            {
                if (record_right_ptr < len_arr)
                {
                    record_left_diff = (x - arr[record_left_ptr]); //Record left-half difference
                    record_right_diff = (arr[record_right_ptr] - x); //Record right-half difference

                    ///// Check if the current difference matched conditions or not /////
                    if (record_left_diff <= record_right_diff) 
                    { 
                        res_vect.insert((res_vect.begin() + 0), arr[record_left_ptr]); //Keep updating, recording

                        (record_left_ptr--); //Keep updating, reducing
                    }

                    else
                    {
                        res_vect.emplace_back(arr[record_right_ptr]); //Keep updating, recording

                        (record_right_ptr++); //Keep updating, accumulating
                    }
                }

                else
                {
                    res_vect.insert((res_vect.begin() + 0), arr[record_left_ptr]); //Keep updating, recording

                    (record_left_ptr--); //Keep updating, reducing
                }
            }

            else
            {
                if (record_right_ptr < len_arr)
                {
                    res_vect.emplace_back(arr[record_right_ptr]); //Keep updating, recording

                    (record_right_ptr++); //Keep updating, accumulating
                }

                else { ; }
            }

        } //K-th times

        return res_vect;
    }
};
